README


This archive contains the distribution Net-Async-Statsd,
version 0.004:

  IO::Async support for statsd/graphite

This software is copyright (c) 2014 by Tom Molesworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Net-Async-Statsd

0.004     2014-08-29 06:16:01+01:00 Europe/London
	New features:
	* Support for prefix option in client

0.003     2014-08-27 03:50:57+01:00 Europe/London
	No new features

	Bugs fixed:
	* Retain the queue_stat Future to avoid ->then in void context,
	as a precaution for when UDP ->send gains the ability to resolve the
	future on completion

	Dependency fixes:
	* sum0 is a relatively new addition to List::Util

0.002     2014-08-09 18:25:21+01:00 Europe/London
	No new features

	Dependency fixes:
	* Added IO::Socket::IP to cpanfile

0.001     2014-08-04 15:44:06+01:00 Europe/London
	Initial CPAN release


lib/Net/Async/Statsd.pm
package Net::Async::Statsd;
# ABSTRACT: IO::Async support for statsd/graphite
use strict;
use warnings;

our $VERSION = '0.004';

=head1 NAME

Net::Async::Statsd - asynchronous API for Etsy's statsd protocol

=head1 VERSION

Version 0.004

=head1 SYNOPSIS

 use Future;
 use IO::Async::Loop;
 use Net::Async::Statsd::Client;
 my $loop = IO::Async::Loop->new;
 $loop->add(my $statsd = Net::Async::Statsd::Client->new(
   host => 'localhost',
   port => 3001,
 ));
 Future->needs_all(
  $statsd->timing(
   'some.task' => 133,
  ),
  $statsd->gauge(
   'some.value' => 80,
  )
 )->get;

=head1 DESCRIPTION

Provides an asynchronous API for statsd.

You probably wanted the client implementation - see L<Net::Async::Statsd::Client>.
There's a basic server implementation in L<Net::Async::Statsd::Server>, note that
this does little more than accept traffic and raise events.

If you're not using L<IO::Async>, this module is probably not what you wanted.
See L<Net::Statsd> instead.

=cut

use Net::Async::Statsd::Server;
use Net::Async::Statsd::Client;

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<Net::Statsd> - synchronous implementation

=back

=head1 AUTHOR

Tom Molesworth <cpan@perlsite.co.uk>

=head1 LICENSE

Copyright Tom Molesworth 2014. Licensed under the same terms as Perl itself.


lib/Net/Async/Statsd/Server.pm
package Net::Async::Statsd::Server;
$Net::Async::Statsd::Server::VERSION = '0.004';
use strict;
use warnings;

use parent qw(IO::Async::Notifier);

=head1 NAME

Net::Async::Statsd::Server - asynchronous server for Etsy's statsd protocol

=head1 VERSION

Version 0.004

=head1 SYNOPSIS

 use Future;
 use IO::Async::Loop;
 use Net::Async::Statsd::Server;
 my $loop = IO::Async::Loop->new;
 $loop->add(my $statsd = Net::Async::Statsd::Server->new(
   port => 3001,
 ));
 $statsd->bus->subscribe_to_event(
  count => sub {
   my ($ev, $k, $delta, $type) = @_;
  }
 );

=head1 DESCRIPTION

Provides an asynchronous server for the statsd API.

=cut

use curry;
use Socket qw(SOCK_DGRAM);
use IO::Socket::IP;
use IO::Async::Socket;

use Net::Async::Statsd::Bus;

=head1 METHODS

All public methods return a L<Future> indicating when the write has completed.
Since writes are UDP packets, there is no guarantee that the remote will
receive the value, so this is mostly intended as a way to detect when
statsd writes are slow.

=cut

=head2 host

Which host to listen on. Probably want '0.0.0.0' (set via L</configure>)
here if you want to listen on all addresses.

=cut

sub host { shift->{host} }

=head2 port

The UDP port we'll accept traffic on. Use L</configure> to set it.

=cut

sub port { shift->{port} }

=head2 configure

Used for setting values.

=cut

sub configure {
	my ($self, %args) = @_;
	for (qw(port host)) {
		$self->{$_} = delete $args{$_} if exists $args{$_};
	}
	$self->SUPER::configure(%args);
}

=head2 listening

Resolves with the port number when the UDP server is listening.

=cut

sub listening {
	my ($self) = @_;
	$self->{listening} ||= do {
		$self->listen
	}
}

=head2 listen

Establishes the underlying UDP socket.

=cut

sub listen {
	my ($self) = @_;

	my $f = $self->loop->new_future;
	my $sock = IO::Socket::IP->new(
		Proto     => 'udp',
		ReuseAddr => 1,
		Type      => SOCK_DGRAM,
		LocalPort => $self->port // 0,
		Listen    => $self->listen_backlog,
		Blocking  => 0,
	) or die "No bind: $@\n";
	$self->{port} = $sock->sockport;
	my $ias = IO::Async::Socket->new(
		handle        => $sock,
		on_recv       => $self->curry::on_recv,
		on_recv_error => $self->curry::on_recv_error,
	);
	$self->add_child($ias);
	$f->done($self->port);
}

=head2 bus

Returns the L<Net::Async::Statsd::Bus> instance for this server.

This object exists purely for the purpose of dispatching events.

=cut

sub bus { shift->{bus} ||= Net::Async::Statsd::Bus->new }

=head2 listen_backlog

Default listen backlog. Immutable, set to 4096 for no particular reason.

=cut

sub listen_backlog { 4096 }

{
my %type = (
	ms => 'timing',
	c => 'count',
	g => 'gauge',
);

=head2 type_for_char

Badly-named lookup method - returns the type matching the given characters.

=cut

sub type_for_char {
	my ($self, $char) = @_;
	die "no character?" unless defined $char;
	return $type{$char};
}
}

=head2 on_recv

Called if we receive data.

=cut

sub on_recv {
	my ($self, undef, $dgram, $addr) = @_;
	$self->loop->resolver->getnameinfo(
		addr    => $addr,
		numeric => 1,
		dgram   => 1,
	)->on_done(sub {
		my ($host, $port) = @_;
		$self->debug_printf("UDP packet received from %s", join ':', $host, $port);
		my ($k, $v, $type_char, $rate) = $dgram =~ /^([^:]+):([^|]+)\|([^|]+)(?:\|\@(.+))?$/ or warn "Invalid dgram: $dgram";
		$rate ||= 1;
		my $type = $self->type_for_char($type_char) // 'unknown';
		$self->bus->invoke_event(
			$type => ($k, $v, $rate, $host, $port)
		);
		$self->debug_printf(
			"dgram %s from %s: %s => %s (%s)",
			$dgram,
			join(':', $host, $port),
			$k,
			$v,
			$type
		);
	});
}

=head2 on_recv_error

Called if we had an error while receiving.

=cut

sub on_recv_error {
	my ($self, undef, $err) = @_;
	$self->debug_printf("UDP packet receive error: %s", $err);
}

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<Net::Statsd> - synchronous implementation

=back

=head1 AUTHOR

Tom Molesworth <cpan@perlsite.co.uk>

=head1 LICENSE

Copyright Tom Molesworth 2014. Licensed under the same terms as Perl itself.


lib/Net/Async/Statsd/Client.pm
package Net::Async::Statsd::Client;
$Net::Async::Statsd::Client::VERSION = '0.004';
use strict;
use warnings;

use parent qw(IO::Async::Notifier);

=head1 NAME

Net::Async::Statsd::Client - asynchronous API for Etsy's statsd protocol

=head1 VERSION

Version 0.004

=head1 SYNOPSIS

 use Future;
 use IO::Async::Loop;
 use Net::Async::Statsd::Client;
 my $loop = IO::Async::Loop->new;
 $loop->add(my $statsd = Net::Async::Statsd::Client->new(
   host => 'localhost',
   port => 3001,
 ));
 # Wait until the stats are written before proceeding
 Future->needs_all(
  $statsd->timing(
   'some.task' => 133,
  ),
  $statsd->gauge(
   'some.value' => 80,
  )
 )->get;
 # Fire-and-forget stat, record 25% of the time:
 $statsd->increment('startup', 0.25);

=head1 DESCRIPTION

Provides an asynchronous API for statsd.

=head1 METHODS

All public methods return a L<Future> indicating when the write has completed.
Since writes are UDP packets, there is no guarantee that the remote will
receive the value, so this is mostly intended as a way to detect when
statsd writes are slow.

=cut

=head2 timing

Records timing information in milliseconds. Takes up to three parameters:

=over 4

=item * $k - the statsd key

=item * $v - the elapsed time in milliseconds

=item * $rate - optional sampling rate

=back

Only the integer part of the elapsed time will be sent.

Example usage:

 $statsd->timing('some.key' => $ms, 0.1); # record this 10% of the time

Returns a L<Future> which will be resolved when the write completes.

=cut

sub timing {
	my ($self, $k, $v, $rate) = @_;

	$self->queue_stat(
		$k => int($v) . '|ms',
		$rate
	);
}

=head2 gauge

Records a current value. Takes up to three parameters:

=over 4

=item * $k - the statsd key

=item * $v - the new value

=item * $rate - optional sampling rate

=back

Only the integer value will be sent.

Example usage:

 $statsd->timing('some.key' => 123);

Returns a L<Future> which will be resolved when the write completes.

=cut

sub gauge {
	my ($self, $k, $v, $rate) = @_;

	$self->queue_stat(
		$k => int($v) . '|g',
		$rate
	);
}

=head2 delta

Records changed value. Takes up to three parameters:

=over 4

=item * $k - the statsd key

=item * $v - the change (positive or negative)

=item * $rate - optional sampling rate

=back

Values are truncated to integers.

Example usage:

 $statsd->timing('some.key' => -12);

Returns a L<Future> which will be resolved when the write completes.

=cut

sub delta {
	my ($self, $k, $v, $rate) = @_;

	$self->queue_stat(
		$k => int($v) . '|c',
		$rate
	);
}

=head2 count

Alias for L</delta>.

=cut

# an alias for good measure
*count = *delta;

=head2 increment

Shortcut for L</delta> with a value of +1.

=cut

sub increment {
	my ($self, $k, $rate) = @_;

	$self->queue_stat(
		$k => '1|c',
		$rate
	);
}

=head2 decrement

Shortcut for L</delta> with a value of -1.

=cut

sub decrement {
	my ($self, $k, $rate) = @_;

	$self->queue_stat(
		$k => '-1|c',
		$rate
	);
}

=head2 configure

Standard L<IO::Async::Notifier> configuration - called on construction or
manually when values need updating.

Accepts the following named parameters:

=over 4

=item * host - the host we'll connect to

=item * port - the UDP port to send messages to

=item * default_rate - default sampling rate when none is provided for a given call

=item * prefix - string to prepend to any stats we record

=back

=cut

sub configure {
	my ($self, %args) = @_;
	for (qw(port host default_rate prefix)) {
		$self->{$_} = delete $args{$_} if exists $args{$_};
	}
	$self->SUPER::configure(%args);
}

=head1 INTERNAL METHODS

These methods are used internally, and are documented
for completeness. They may be of use when subclassing
this module.

=cut

=head2 queue_stat

Queues a statistic for write.

=cut

sub queue_stat {
	my ($self, $k, $v, $rate) = @_;

	$rate //= $self->default_rate;
	return Future->wrap unless $self->sample($rate);

	$k = $self->{prefix} . '.' . $k if exists $self->{prefix};

	# Append rate if we're only sampling part of the data
	$v .= '|@' . $rate if $rate < 1;
	my $f;
	$f = $self->statsd->then(sub {
		# FIXME Someday IO::Async::Socket may support
		# Futures for UDP send, update this if/when
		# that happens.
		shift->send("$k:$v");
		Future->wrap
	})->on_ready(sub { undef $f });
}

=head2 sample

Applies sampling based on the given rate - returns true if
we should record this, false otherwise.

=cut

sub sample {
	my ($self, $rate) = @_;
	return 1 if rand() <= $rate;
	return 0;
}

=head2 default_rate

Default sampling rate. Currently 1 if not overidden in constructor or L</configure>.

=cut

sub default_rate { shift->{default_rate} // 1 }

=head2 port

Statsd UDP port.

=cut

sub port { shift->{port} }

=head2 host

Statsd host to connect to.
=cut

sub host { shift->{host} }

sub statsd {
	my ($self) = @_;
	$self->{statsd} ||= do {
		$self->connect
	}
}

=head2 connect

Establishes the underlying UDP socket.

=cut

sub connect {
	my ($self) = @_;
	# IO::Async::Loop
	$self->loop->connect(
		family    => 'inet',
		socktype  => 'dgram',
		service   => $self->port,
		host      => $self->host,
		on_socket => $self->curry::on_socket,
	);
}

=head2 on_socket

Called when the socket is established.

=cut

sub on_socket {
	my ($self, $sock) = @_;
	$self->debug_printf("UDP socket established: %s", $sock->write_handle->sockhost_service);
	# FIXME Don't really want this - we're sending only, no bi-directional shenanigans
	# required, might need to replace ->connect with an IO::Async::Socket for this?
	$sock->configure(
		on_recv       => $self->curry::weak::on_recv,
		on_recv_error => $self->curry::weak::on_recv_error,
	);
	$self->add_child($sock);
}

=head2 on_recv

Called if we receive data.

=cut

sub on_recv {
	my ($self, undef, $dgram, $addr) = @_;
	$self->debug_printf("UDP packet [%s] received from %s", $dgram, join ':', $self->loop->resolver->getnameinfo(
		addr    => $addr,
		numeric => 1,
		dgram   => 1,
	));
}

=head2 on_recv_error

Called if we had an error while receiving.

=cut

sub on_recv_error {
	my ($self, undef, $err) = @_;
	$self->debug_printf("UDP packet receive error: %s", $err);
}

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<Net::Statsd> - synchronous implementation

=back

=head1 AUTHOR

Tom Molesworth <cpan@perlsite.co.uk>

=head1 LICENSE

Copyright Tom Molesworth 2014. Licensed under the same terms as Perl itself.
